stdlib/math.pyi

```diff
@@ -130,6 +133,10 @@ def radians(__x: _SupportsFloatOrIndex) -> float: ...
 def remainder(__x: _SupportsFloatOrIndex, __y: _SupportsFloatOrIndex) -> float: ...
 def sin(__x: _SupportsFloatOrIndex) -> float: ...
 def sinh(__x: _SupportsFloatOrIndex) -> float: ...
+
+if sys.version_info >= (3, 12):
+    def sumprod(__p: Iterable[_SupportsFloatOrIndex], __q: Iterable[_SupportsFloatOrIndex]) -> float: ...
+
```

Comment:
Seems like you need types that support multiplication:

```pycon
>>> import math as m
>>> class Floatable:
...     def __float__(self):
...         return 42.0
...
>>> class Indexable:
...     def __index__(self):
...         return 42
...
>>> indexes = [Indexable(), Indexable()]
>>> floats = [Floatable(), Floatable()]
>>> m.sumprod(indexes, indexes)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'Indexable' and 'Indexable'
>>> m.sumprod(floats, floats)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'Floatable' and 'Floatable'
>>> m.sumprod(indexes, floats)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'Indexable' and 'Floatable'
>>> m.sumprod(floats, indexes)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'Floatable' and 'Indexable'
```

Reply:
Rather than make a mul and add protocol where I'm not quite sure of the parameter types, I opted to just go for float. If you're using e.g. numpy types you're not going to be using math.sumprod anyway.

Reply:
Yeah, good call, I think — BinOp protocols are always annoyingly complex to get right anyway